pkg/agent/manager/cache/cache.go

```diff
@@ -107,6 +107,8 @@ type Cache struct {
 
 	mu sync.RWMutex
 
+	mux sync.Mutex
```

Comment:
we already have the `mu` usage in many places, which ultimately wind up calling `getSelectorIndex`

It might be worth instead tracking down where is the originating call to `getSelectorIndex` that *does not* use `mu`.

Reply:
I haven't audited myself, but I think the more likely scenario is that the selector index is being mutated while only the read lock is held, allowing concurrent readers to access it during the same time. This is an oversight, wherein the selector index mutation should take place under a write lock. To fix this, we'll need to rework how the selector index is populated. However, I think we can get by without adding another mutex.

Reply:
Yup, found it. `FetchWorkloadUpdate` only takes a read-lock and indirectly calls getSelectorIndex. Two solutions come to mind:
1) Change `FetchWorkloadUpdate` to take a write lock. This would probably work fine since this call is very fast, but isn't ideal. On this call, since there is no subscription happening, it shouldn't populate the selector index anyway.
2) Have a read-only mode variant of getSelectorIndex that doesn't always add in the selector index for the selector that is passed in. This is more _correct_. It will be a little more involved to change though.

Reply:
Number 2 is actually pretty simple.

Reply:
> Yup, found it. `FetchWorkloadUpdate` only takes a read-lock and indirectly calls getSelectorIndex. Two solutions come to mind:
> 
> 1. Change `FetchWorkloadUpdate` to take a write lock. This would probably work fine since this call is very fast, but isn't ideal. On this call, since there is no subscription happening, it shouldn't populate the selector index anyway.
> 2. Have a read-only mode variant of getSelectorIndex that doesn't always add in the selector index for the selector that is passed in. This is more _correct_. It will be a little more involved to change though.

Glad to hear from you guys, I don't think `FetchWorkloadUpdate` is the only one at risk, we trigger the exception because `FetchJWTSVID` calls `MatchingIdentities` which triggers this issue, so I think number 1 will be a bit more complicated and may degrade the performance of the workload fetching jwtsvid.

Number 2 I didn't think of an easy enough way to modify it, that's why I added a new Mutex, so hope to get your guidance.

Reply:
> Yup, found it. `FetchWorkloadUpdate` only takes a read-lock and indirectly calls getSelectorIndex. Two solutions come to mind:
> 
> 1. Change `FetchWorkloadUpdate` to take a write lock. This would probably work fine since this call is very fast, but isn't ideal. On this call, since there is no subscription happening, it shouldn't populate the selector index anyway.
> 2. Have a read-only mode variant of getSelectorIndex that doesn't always add in the selector index for the selector that is passed in. This is more _correct_. It will be a little more involved to change though.

@azdagron @amoore877 Glad to hear from you guys, I don't think `FetchWorkloadUpdate` is the only one at risk, we trigger the exception because `FetchJWTSVID` calls `MatchingIdentities` which triggers this issue, so I think number 1 will be a bit more complicated and may degrade the performance of the workload fetching jwtsvid.

Number 2 I didn't think of an easy enough way to modify it, that's why I added a new Mutex, so hope to get your guidance.

Reply:
Yes! There are other callers, like you mention. I think number 1 is off the table.

I don't expect you to do the extra work for number 2 :) I've gone ahead and made the changes locally. I'm happy to push them up to this PR or a clean PR. If this PR, I'll need you to fix DCO on your first commit.

Reply:
@loveyana even if this specific PR is unused, reporting the discovery, root causing, and submitting a fix are still very much appreciated :) Looking forward to future contributions

Reply:
What @amoore877 said :) Thank you very much.